blocks/algorithms/__init__.py

```diff
@@ -244,12 +269,66 @@ class SteepestDescent(StepRule):
 
     Attributes
     ----------
-    learning_rate : float
-        The learning rate.
+    learning_rate : :class:`~tensor.TensorSharedVariable`
+        The shared variable storing the learning rate used.
 
     """
     def __init__(self, learning_rate=1.0):
-        self.learning_rate = learning_rate
+        self.learning_rate = shared_floatx(learning_rate)
+
+    def compute_step(self, param, gradient):
+        return -self.learning_rate * gradient
+
+
+class GradientClipping(StepRule):
+    """Clips the total gradient to make it not exceed a threshold.
 
-    def compute_step(self, param, grad_wr_param):
-        return -self.learning_rate * grad_wr_param
+    Parameters
+    ----------
+    threshold : float, optional
+        The maximum permitted L2 norm for the gradient. The gradient
+        will be rescaled to be not higher than this quanity.
+        If ``None``, no rescaling will be applied.
```

Comment:
Fair enough. Regarding hyperparameter optimization, @dwf do you know if hyperopt is any good, and worth looking into integrating eventually? I have to say that I haven't given hyperparameter searches a lot of thought, but we'll need to make sure it's easy enough to do.